lean/main/dsls.lean

```diff
@@ -0,0 +1,514 @@
+> 2. Definitionally equality checking with support for metavariables (aka unification modulo definitional equality).
+> 3. Type inference.
+> 4. Type class resolution.
+
+We do not explore the full power of `MetaM` in this tutorial, and simply 
+gesture at how to get access to this low-level machinery.
+
+Concretely, we teach how to assemble
+elaborated pieces of Lean4 with the goal of providing an accessible syntax to express
+to the [IMP](http://concrete-semantics.org/concrete-semantics.pdf) DSL,
+which is a simple imperative language, often used for teaching operational
+and denotational semantics. 
+
+
+
+## DSLs in Lean via TermElabM
+
+Let's begin with the usual incantations, where we import `Lean` and open
+`Lean`, `Lean.Elab`, and `Lean.Meta`.
+
+-/
+import Lean
+import Lean.Elab.Deriving.Basic
+open Lean Lean.Syntax
+open Lean.Elab
+open Lean.Elab.Command
+open Lean.Meta
+open Lean
+
+/--
+We begin by defining the AST of the language:
+We have arithmetic expressions, boolean expressions, and commands.
+-/
+
+/-
+#### Arithmetic Expressions
+
+Arithmetic expressions are naturals, variables, or sums of other
+arithmetic expressions.
+-/
+inductive AExp
+| ANat: Nat -> AExp
+| AVar: String -> AExp
+| APlus: AExp -> AExp -> AExp
+deriving Inhabited
+
+/-
+#### Boolean Expressions
+
+Boolean expressions are either booleans (true/false), variables,
+an and (`&&`) of two booleans, or a less than comparison (`<`) between
+two arithmetic expressions.
+-/
+inductive BExp
+| BBool: Bool -> BExp
+| BVar: String -> BExp
+| BAnd: BExp -> BExp -> BExp
+| BLess: AExp -> AExp -> BExp
+deriving Inhabited
+
+/-
+#### Command
+
+Commands can be either a `skip` command to skip the execution, an `assign`
+command to assign a value to a variable, a `seq`(`;;`) to sequence commands,
+an `if` for conditionals, and `while` for looping.
+-/
+inductive Command
+| Skip: Command
+| Assign: String -> AExp -> Command
+| Seq: Command -> Command -> Command
+| If: BExp -> Command -> Command -> Command
+| While: BExp -> Command -> Command
+deriving Inhabited
+
+
+/-
+## Embedding a DSL via Low-level Syntax Elaboration
+
+In this section, we shall contrast the previously explained
+macro-based approach (which has type `Syntax → MacroM Syntax`)
+with the lower level elaborator (which has type `Syntax → TermElabM Expr`)
+that is the focus of this section.
+-/
+
+/-
+#### Parsing AExp via Syntax Elaboration
+-/
+
+
+declare_syntax_cat imp_aexp
+syntax num : imp_aexp
+syntax ident : imp_aexp
+syntax imp_aexp "+" imp_aexp : imp_aexp
+
+/- 
+Recall that if we were doing this via lean macros, 
+we would write out an "interpretation" macro such as `[imp_aexp|...]`,
+and the declare the translation as a macro rule:
+-/
+
+syntax "[imp_aexp|" imp_aexp "]" : term
+macro_rules
+| `([imp_aexp| $n:num ]) => do
+   let n : Nat := n.toNat -- grab the number.
+   let nStx : Syntax := Lean.quote n -- quote the number.
+   `(AExp.ANat $(nStx)) -- return the syntax.
```

Comment:
This should work, right?
```suggestion
| `([imp_aexp| $n:num ]) => `(AExp.ANat n)
```

Reply:
Ah yes, that makes much more sense, than to pull out data and then quote again :D